bootstrap/package_manager.py

```diff
@@ -0,0 +1,259 @@
+        lines = [line.rsplit(':', 1)[0] for line in lines]
+        return set(lines)
+
+    def install_compiler(self) -> str:
+        if distro.id() == "ubuntu" and int(distro.major_version()) < 22:
+            self.install({"compiler_prereq": {"apt-transport-https", "ca-certificates", "software-properties-common"}})
+            _run_command_with_confirm("sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test",
+                                      no_confirm=self.no_confirm)
+            self.install({"compiler": {"build-essential", "g++-11"}})
+            return "-DCMAKE_C_COMPILER=gcc-11 -DCMAKE_CXX_COMPILER=g++-11"
+        self.install({"compiler": {"g++"}})
+        return ""
+
+
+class DnfPackageManager(PackageManager):
+    def __init__(self, no_confirm):
+        PackageManager.__init__(self, no_confirm)
+
+    def install(self, dependencies: Dict[str, Set[str]]) -> bool:
+        return self._install(dependencies=dependencies,
+                             install_cmd="sudo dnf --enablerepo=crb install -y epel-release",
+                             replace_dict={"gpsd": {"gpsd-devel"},
+                                           "libpcap": {"libpcap-devel"},
+                                           "lua": {"lua-devel"},
+                                           "python": {"python3-devel"},
+                                           "jni": {"java-1.8.0-openjdk", "java-1.8.0-openjdk-devel", "maven"},
+                                           "libpng": {"libpng-devel"},
+                                           "libusb": {"libusb-devel"}})
+
+    def _get_installed_packages(self) -> Set[str]:
+        result = subprocess.run(['dnf', 'list', 'installed'], text=True, capture_output=True, check=True)
+        lines = [line.split(' ')[0] for line in result.stdout.splitlines()]
+        lines = [line.rsplit('.', 1)[0] for line in lines]
+        return set(lines)
+
+    def install_compiler(self) -> str:
+        self.install({"compiler": {"gcc-c++"}})
+        return ""
+
+
+class PacmanPackageManager(PackageManager):
+    def __init__(self, no_confirm):
+        PackageManager.__init__(self, no_confirm)
+
+    def install(self, dependencies: Dict[str, Set[str]]) -> bool:
+        return self._install(dependencies=dependencies,
+                             install_cmd="sudo pacman --noconfirm -S",
+                             replace_dict={"jni": {"jdk8-openjdk", "maven"}})
+
+    def _get_installed_packages(self) -> Set[str]:
+        result = subprocess.run(['pacman', '-Qq'], text=True, capture_output=True, check=True)
+        return set(result.stdout.splitlines())
+
+    def install_compiler(self) -> str:
+        self.install({"compiler": {"gcc"}})
+        return ""
+
+
+def _get_vs_dev_cmd_path() -> str:
+    vswhere_results = subprocess.run(
+        "vswhere -products * -property installationPath -requires Microsoft.VisualStudio.Component.VC.ATL",
+        capture_output=True)
```

Comment:
`vswhere` is not found on my machine. It's present in the VS installer directory, but not in the path. When I installed it with scoop, this could use that version, but not the one that's automatically installed with visual studio. 
![image](https://github.com/apache/nifi-minifi-cpp/assets/1170582/f44142ea-308e-4710-94d0-7bb39d2b8ebe)


This script is supposed to work even without VS installed, right? Where does it take vswhere from in that case?

Reply:
Yeah the script uses the package manager to tell if a required package/lib/exe is installed or not.
Maybe instead of the package manager we could use where.exe to tell if vswhere is installed or not. Probably this is the only instance where this would be useful (and vswhere is a tiny exe), so not sure if it worth the hassle to add more complication to the code. What do you think?

Reply:
I think it's worth having an exception for vswhere.exe if it's already installed with VS. According to microsoft/vswhere readme, it's always at `%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe` when it's installed.
There's also a problem with it finding VS2019 instead of VS2022, as I have both installed to be able to compile older versions of minifi.

Reply:
I've addressed these issues in https://github.com/apache/nifi-minifi-cpp/pull/1681/commits/46b3ceb968812ee09fdd6eca70aba37b967a2f25